baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,511 @@
+
+        Parameters
+        ----------
+        clearMetadata : bool, optional
+            Whether to clear the metadata too, by default False
+        """
+
+        # Reset iteration counter
+        self._iter = 0
+
+        # Reset solve start time
+        self._startTime = -1.0
+
+        # Clear recorded data
+        for variable in self._variables.values():
+            variable.reset()
+
+        # Clear metadata if required
+        if clearMetadata:
+            self._metadata = {}
+
+    def addMetadata(self, name: str, data: Any) -> None:
+        """Add a piece of metadata to the history
+
+        The metadata attribute is simply a dictionary that can be used to store arbitrary information related to the
+        solution being recorded, e.g solver options
+
+        Parameters
+        ----------
+        name : str
+            Item name/key
+        data : Any
+            Item to store
+        """
+        self._metadata[name] = data
+
+    def addVariable(
+        self,
+        name: str,
+        varType: Type,
+        printVar: bool = False,
+        valueFormat: Optional[str] = None,
+        overwrite: bool = False,
+    ) -> None:
+        """Define a new field to be stored in the history.
+
+        Parameters
+        ----------
+        name : str
+            Variable name
+        varType : Type
+            Variable type, i.e int, float, str etc
+        printVar : bool, optional
+            Whether to include the variable in the iteration printout, by default False
+        valueFormat : str, optional
+            Format string valid for use with the str.format() method (e.g "{:17.11e}" for a float or "{:03d}" for an
+            int), only important for variables that are to be printed. By default a predefined format for the given
+            `varType` is used
+        overwrite : bool, optional
+            Whether to overwrite any existing variables with the same name, by default False
+        """
+
+        if name not in self._variables or overwrite:
+
+            # Only store variable's string format if it's going to be printed
+            if not printVar:
+                valueFormat = None
+                headerFormat = None
+            else:
+                if valueFormat is not None:
+                    pass
+                elif varType in self._defaultFormat:
+                    valueFormat = self._defaultFormat[varType]
+                else:
+                    valueFormat = self._DEFAULT_OTHER_FORMAT
+
+                # Get test value for figuring out how long a string is using the supplied format
+                try:
+                    testValue = self._testValues[varType]
+                except KeyError:
+                    testValue = self._DEFAULT_OTHER_VALUE
+
+                # Figure out column width, the maximum of the length of the name and the formatted value, also check
+                # that the format string is valid
+                try:
```

Comment:
@eirikurj I wanted to get rid of `self._testValues` now that your format parser is merged, but I can't think of a way to test if the supplied format is valid without having a value to test it on. Do you have any better ideas?

Reply:
Dont think its easy to test the general validity of a format string without a value. The user can always supply some format string and the value can be a garbage or different type.
Maybe one possibility without a value is to check from the parsed format that `ftype` and the supplied datatype match.

Reply:
Solved by using the value returned by the default constructor for the provided type (i.e `varType()`)